File: src/main/java/com/raiderrobotix/_2020/commands/drivebase/Turn.java
package com.raiderrobotix._2020.commands.drivebase;

import static com.raiderrobotix._2020.subsystems.DriveBase.drives;

import com.raiderrobotix._2020.commands.Command;

public class Turn extends Command {

	private static final double TURN_TOLERANCE = 1.0; // TODO

	private final double angle;
	private final double speed;

	public Turn(double angle, double speed) {
		requires(drives);
		this.angle = angle;
		this.speed = Math.copySign(speed, angle);
	}

	@Override
	protected void initialize() {
		drives.resetGyro();
	}

	@Override
	protected void execute() {
		drives.setSpeed(speed, -speed);
	}

	@Override
	protected void end() {
		drives.setSpeed(0);
	}

	@Override
	protected boolean isFinished() {
		return Math.abs(drives.getGyroAngle() - angle) < TURN_TOLERANCE;
	}

}

File: src/main/java/com/raiderrobotix/_2020/commands/drivebase/Turn.kt
package com.raiderrobotix._2020.commands.drivebase

import com.raiderrobotix._2020.commands.Command
import com.raiderrobotix._2020.subsystems.DriveBase
import kotlin.math.abs

/**
 * @param speed must positive
 */
class Turn(private val angle: Double, private val speed: Double) : Command() {
	
	private val TURN_TOLERANCE = 1.0 // TODO
	
	init {
		requires(DriveBase)
	}
	
	override fun initialize() {
		DriveBase.resetGyro()
	}
	
	override fun execute() {
		DriveBase.setSpeed(speed, -speed)
	}
	
	override fun end() {
		DriveBase.speed = 0.0
	}
	
	override fun isFinished(): Boolean  = abs(DriveBase.gyroAngle - angle) < TURN_TOLERANCE
	
}